src/core/batch.rkt

```diff
@@ -129,11 +132,50 @@
                [args args])
       (define idx (batchref-idx brf))
       (cond
-        [(and (> (dvector-capacity visited) idx) (dvector-ref visited idx)) (dvector-ref out idx)]
+        [(and (> (dvector-capacity visited) idx) (dvector-ref visited idx))
+         (when (or (and (null? args)
+                        (not (equal? #t (dvector-ref visited idx)))) ;; fast-check for null-arguments
+                   (and (not (null? args))
+                        (not (equal? args
+                                     (dvector-ref visited idx))))) ;; check for non-null arguments
+           (error 'batch-recurse "Cache violation with a different argument ~a for ~a" args brf))
+         (dvector-ref out idx)]
         [else
          (define res (apply f brf (λ (brf . args) (loop brf args)) args))
          (dvector-set! out idx res)
-         (dvector-set! visited idx #t)
+         (if (null? args) ;; updating cache
+             (dvector-set! visited idx #t) ;; fast-caching
+             (dvector-set! visited idx args))
```

Comment:
I don't think this is any faster. The empty list is a singleton (there's only one, it's always `eq?`) so there's no benefit to this.